### PR TITLE
luminous: osd: may get empty info at recovery

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/24772

### CODE REVIEW

**Comment by smithfarm** [comment]
@liewegas See the conflict resolution statement in the second commit message. Could https://github.com/ceph/ceph/commit/c1fbf350bcafab0ad94306302217eb39a66814b7 be the reason why this bug did not happen pre-Mimic?

**Comment by smithfarm** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/PG.cc: In member function ‘bool PG::search_for_missing(const pg_info_t&, const pg_missing_t&, pg_shard_t, PG::RecoveryCtx*)’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/PG.cc:566:12: error: expected ‘;’ before ‘.’ token
       oinfo.last_update != eversion_t()) {
            ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/PG.cc:565:41: warning: statement has no effect [-Wunused-value]
        CEPH_FEATURE_OSD_ERASURE_CODES)) &&
                                         ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/PG.cc:566:7: error: label ‘oinfo’ used but not defined
       oinfo.last_update != eversion_t()) {
       ^
src/osd/CMakeFiles/osd.dir/build.make:134: recipe for target 'src/osd/CMakeFiles/osd.dir/PG.cc.o' failed
```

File: src/osd/PG.cc [hunk 1]
```diff
@@ -556,9 +556,14 @@ bool PG::search_for_missing(
...
     publish_stats_to_osd();
+  // avoid doing this if the peer is empty.  This is abit of paranoia
+  // to avoid doing something rash if add_source_info() above
+  // incorrectly decided we found something new. (if the peer has
+  // last_update=0'0 that's impossible.)
   if (found_missing &&
       (get_osdmap()->get_features(CEPH_ENTITY_TYPE_OSD, NULL) &
-       CEPH_FEATURE_OSD_ERASURE_CODES)) {
+       CEPH_FEATURE_OSD_ERASURE_CODES) &&
+      oinfo.last_update != eversion_t()) {
```
Comment (smithfarm) [inline src/osd/PG.cc]: Mimic and above do not check for CEPH_FEATURE_OSD_ERASURE_CODES in this conditional - see c1fbf350bcafab0ad94306302217eb39a66814b7

**Comment by liewegas** [comment]
This looks okay, but I would like to let this one sit in master for a little while before we merge the backport to luminous.  Let's target the next point release?

**Comment by smithfarm** [comment]
@liewegas OK

Note to self: DNM can be removed after 12.2.6 is released.

**Comment by yuriw** [comment]
[wip-yuri-testing-2018-08-01-1605-luminous](https://trello.com/c/DxuaJWNO/360-wip-yuri-testing-2018-08-01-1605-luminous)

**Pull request merged**